File: backend/src/modules/gateway/request/local-request-errors.ts
import type { Request, Response } from 'express'

import { logger } from '../../../shared/logger.js'
import type { AuditCaptureContext } from '../audit/capture.service.js'
import {
  recordClientIpErrorCircuitSampleAsync,
  type GatewayClientIpErrorCircuitReason
} from '../runtime/client-ip-error-circuit.service.js'
import { sendGatewayFailureResponse } from '../response/failure-response.js'
import { gatewayErrorPayload } from '../response/responses.js'
import type { GatewayFailureUsageContext } from '../usage/records.js'

export async function sendInvalidJsonGatewayResponse(input: {
  req: Request
  res: Response
  auditCapture: AuditCaptureContext
  usageContext: GatewayFailureUsageContext
  startedAt: number
  systemAccountId: string
  apiKeyId?: string
  groupId: string
  clientIp?: string
  endpoint: string
}): Promise<void> {
  const statusCode = 400
  const responsePayload = gatewayErrorPayload('请求体不是合法 JSON', 'invalid_request_error')
  await recordClientIpRequestErrorSample({
    auditCapture: input.auditCapture,
    systemAccountId: input.systemAccountId,
    apiKeyId: input.apiKeyId,
    groupId: input.groupId,
    clientIp: input.clientIp,
    endpoint: input.endpoint,
    reason: 'invalid_json',
    signature: 'invalid_json'
  })
  await sendGatewayFailureResponse({
    req: input.req,
    res: input.res,
    auditCapture: input.auditCapture,
    usageContext: input.usageContext,
    startedAt: input.startedAt,
    statusCode,
    responsePayload,
    audit: {
      outcome: 'gateway_failed',
      errorPhase: 'request_validation',
      errorCode: 'invalid_json',
      errorMessage: responsePayload.error.message
    }
  })
}

export async function recordClientIpRequestErrorSample(input: {
  auditCapture: AuditCaptureContext
  systemAccountId: string
  apiKeyId?: string
  groupId: string
  clientIp?: string
  endpoint: string
  reason: GatewayClientIpErrorCircuitReason
  signature?: string
}): Promise<void> {
  const result = await recordClientIpErrorCircuitSampleAsync(input)
  if (!result.blocked) {
    return
  }
  logger.warn({
    event: 'gateway_client_ip_error_circuit_opened',
    reason: input.reason,
    retryAfterSeconds: result.retryAfterSeconds,
    failureCount: result.failureCount,
    systemAccountId: input.systemAccountId,
    apiKeyId: input.apiKeyId,
    groupId: input.groupId,
    clientIp: input.clientIp
  }, '客户端 IP 级错误熔断已打开')
  input.auditCapture.addGatewayMetadata({
    label: 'client_ip_error_circuit',
    metadata: {
      opened: true,
      reason: input.reason,
      retryAfterSeconds: result.retryAfterSeconds,
      failureCount: result.failureCount
    }
  })
}

File: backend-go/projects/gateway/internal/gatewaypreauth/localrequesterrors.go
package gatewaypreauth

import (
	"context"
	"net/http"
)

// Port of request/local-request-errors.ts: the local invalid-JSON failure
// response and the client-ip error circuit sampling hook.

// InvalidJSONGatewayResponse mirrors sendInvalidJsonGatewayResponse.
func (s *Service) SendInvalidJSONGatewayResponse(ctx context.Context, input InvalidJSONResponseInput) {
	statusCode := http.StatusBadRequest
	responsePayload := GatewayErrorPayloadOf("请求体不是合法 JSON", "invalid_request_error")
	s.RecordClientIPRequestErrorSample(ctx, input.AuditCapture, ClientIPErrorSampleInput{
		SystemAccountID: input.SystemAccountID,
		APIKeyID:        input.APIKeyID,
		GroupID:         input.GroupID,
		ClientIP:        input.ClientIP,
		Endpoint:        input.Endpoint,
		Reason:          ClientIPErrorCircuitInvalidJSON,
		Signature:       "invalid_json",
	})
	s.Responses.SendGatewayFailureResponse(FailureResponseInput{
		Req:             input.Req,
		Res:             input.Res,
		AuditCapture:    input.AuditCapture,
		UsageContext:    input.UsageContext,
		StartedAt:       input.StartedAt,
		StatusCode:      statusCode,
		ResponsePayload: responsePayload,
		Audit: FailureAudit{
			Outcome:      AuditOutcomeGatewayFailed,
			ErrorPhase:   "request_validation",
			ErrorCode:    "invalid_json",
			ErrorMessage: responsePayload.Error.Message,
		},
	})
}

// InvalidJSONResponseInput mirrors sendInvalidJsonGatewayResponse's input.
type InvalidJSONResponseInput struct {
	Req             *GatewayRequest
	Res             GatewayResponseWriter
	AuditCapture    AuditCaptureContext
	UsageContext    GatewayFailureUsageContext
	StartedAt       int64
	SystemAccountID string
	APIKeyID        string
	GroupID         string
	ClientIP        string
	Endpoint        string
}

// ClientIPErrorSampleInput mirrors recordClientIpRequestErrorSample's input.
type ClientIPErrorSampleInput struct {
	SystemAccountID string
	APIKeyID        string
	GroupID         string
	ClientIP        string
	Endpoint        string
	Reason          ClientIPErrorCircuitReason
	Signature       string
}

// RecordClientIPRequestErrorSample mirrors recordClientIpRequestErrorSample:
// record the sample and surface the opened-circuit diagnostics when the
// sample crosses the threshold.
func (s *Service) RecordClientIPRequestErrorSample(ctx context.Context, auditCapture AuditCaptureContext, input ClientIPErrorSampleInput) {
	result, err := s.Circuits.RecordClientIPErrorCircuitSample(ctx, ClientIPErrorCircuitSampleInput{
		SystemAccountID: input.SystemAccountID,
		APIKeyID:        input.APIKeyID,
		GroupID:         input.GroupID,
		ClientIP:        input.ClientIP,
		Endpoint:        input.Endpoint,
		Reason:          input.Reason,
		Signature:       input.Signature,
	})
	if err != nil || !result.Blocked {
		return
	}
	s.Observability.Logger().Warn("gateway_client_ip_error_circuit_opened", map[string]any{
		"reason":            string(input.Reason),
		"retryAfterSeconds": result.RetryAfterSeconds,
		"failureCount":      result.FailureCount,
		"systemAccountId":   input.SystemAccountID,
		"apiKeyId":          input.APIKeyID,
		"groupId":           input.GroupID,
		"clientIp":          input.ClientIP,
	}, "客户端 IP 级错误熔断已打开")
	auditCapture.AddGatewayMetadata("client_ip_error_circuit", map[string]any{
		"opened":            true,
		"reason":            string(input.Reason),
		"retryAfterSeconds": result.RetryAfterSeconds,
		"failureCount":      result.FailureCount,
	})
}
